crypto/fipsmodule/ec/p521.c

```diff
@@ -495,22 +495,11 @@ static void ec_GFp_nistp521_point_mul(const EC_GROUP *group, EC_JACOBIAN *r,
   p521_felem p_pre_comp[P521_MUL_TABLE_SIZE][3];
 
   // Set the first point in the table to P.
-  p521_from_generic(p_pre_comp[0][0], &p->X);
-  p521_from_generic(p_pre_comp[0][1], &p->Y);
-  p521_from_generic(p_pre_comp[0][2], &p->Z);
+  p521_from_generic(tmp[0], &p->X);
+  p521_from_generic(tmp[1], &p->Y);
+  p521_from_generic(tmp[2], &p->Z);
 
-  // Compute tmp = [2]P.
-  p521_point_double(tmp[0], tmp[1], tmp[2],
-                    p_pre_comp[0][0], p_pre_comp[0][1], p_pre_comp[0][2]);
-
-  // Generate the remaining 15 multiples of P.
-  for (size_t i = 1; i < P521_MUL_TABLE_SIZE; i++) {
-    p521_point_add(p_pre_comp[i][0], p_pre_comp[i][1], p_pre_comp[i][2],
-                   tmp[0], tmp[1], tmp[2], 0 /* both Jacobian */,
-                   p_pre_comp[i - 1][0],
-                   p_pre_comp[i - 1][1],
-                   p_pre_comp[i - 1][2]);
-  }
+  generate_table(p521_methods(), (ec_nistp_felem_limb*)p_pre_comp, tmp[0], tmp[1], tmp[2]);
```

Comment:
Static assert that `P521_MUL_TABLE_SIZE ` is less than `NISTP_FELEM_MAX_NUM_OF_LIMBS`?
This also implicitly rely on contiguous memory for multi-dimensional arrays

Reply:
assert added.

> This also implicitly rely on contiguous memory for multi-dimensional arrays
yes, but I don't think that can be an issue here because the limbs are either 32 or 64 bit wide and a field element is just several limbs, so I don't see a way for the compiler to mess it up. Regardless, I added asserts to make sure that's true.

Reply:
The assert you added should ensure there is enough memory.